### ISSUE TITLE
bug: manual override switch not toggling on

### ISSUE THREAD

**Issue by doppiaemme** [firstof9/openevse#161]
Hi, 
I installed the new UI on OpenEvse but I have the impression the manual override from HA is not working anymore.

The new GUI has three buttons: Enable Charge, Auto, Disable Charge.
When I force it on HA it's not applied anymore and stays on "Auto"

any clue?

thanks!

**Comment by firstof9** [comment]
Seems the scheduler doesn't count for override toggling in the newer firmware, I'll see what I can do.

**Comment by firstof9** [comment]
> do I still have the old openevsehttp?

Looks like you updated it fine, thanks for testing.

I'll work on it some more.

**Comment by firstof9** [comment]
I've dropped a new library, all you need to update is the `manifest.json` file then restart Home Assistant.

**Comment by doppiaemme** [comment]
Hi, 

I am not sure what it is but there are still issues...
I can see at least now it doesn't trigger anymore an error in the log, however turning on and off manual override from HA or from the openevse UI weird things happen.

here for example I don't know how I ended up in this state... manual override on but the charger is sleeping. 
also it seems the setting of "normal" or "eco" mode from HA is not working. 
it updates the HA state when I change it on the wallbox side but not the other way around.

<img width="1066" alt="image" src="https://user-images.githubusercontent.com/3978350/223737439-8b104c82-13b5-4179-b359-223b99ad9a6a.png">
<img width="1252" alt="image" src="https://user-images.githubusercontent.com/3978350/223737594-43575c52-1836-4b6e-8b30-32b811f9d021.png">

**Comment by firstof9** [comment]
Yes "Sleep" actually sends a RAPI command to the EVSE to sleep. I may end up needing to change fully how the override toggle works at this point.

**Comment by fhteagle** [comment]
Agreed that backing up reworking the HA controls is probably the smart play at this point vs trying to fix the existing controls. The GUI is trying to go away from the terms "manual override" and "sleep", in favor of "activate(d)", "auto", and "disable(d)". Would be nice if the HA controls migrated to the same terminology and functionality of the new GUI. Or go one step further than the GUI towards simplification and just make the HA select.openevsename_max_current dropdown have a "0" amps item that disables charging. 

Let me know what I can do to help think it through, design it, test it, etc. Thanks.

**Comment by firstof9** [comment]
The manual override's purpose is now to indicate if you have an override enabled and allow you to toggle it on/off.
Please give `2.1.11` a try.

**Comment by fhteagle** [comment]
Testing 2.1.11 against a Wifi firmware 4.1.8 dev branch openevse unit:

With switch.openevse_manual_override in on position, and switch.openevse_sleep_mode in off position (and matching the GUI and endpoints values), issuing service call to switch: turn_on to entity switch.openevse_sleep_mode results in both switch.openevse_sleep_mode and switch.openevse_manual_override switches turned off. 

Issuing a second service call to switch: turn_on to entity switch.openevse_sleep_mode produces my expected result of the first service call, in that both switch.openevse_manual_override and switch.openevse_sleep_mode end up in the on state.

This behavior is reproduceable against both of my units, its not the device_id issue noted elsewhere.

**Comment by firstof9** [comment]
Expected behavior, no override setup, nothing for it to toggle settings wise.

**Comment by fhteagle** [comment]
I disagree that this is expected behavior. I specifically asked HA to set sleep mode to on. What I get instead is the existing override is cleared and unit is active/not sleeping.

**Comment by doppiaemme** [comment]
[](url)

**Comment by firstof9** [comment]
Setting the max current is an override now.
The override toggle does exactly as described on the tin, toggles override on/off (if set).

**Comment by fhteagle** [comment]
Here's what I am getting when testing 2.1.14 against 8.2.2+4.1.7-dev firmwares on the unit:


https://user-images.githubusercontent.com/9244278/225943065-2fdfded3-f04c-449b-b0f2-0d65c7cd1c2a.mp4



Unexpected actions / results:

1. Clicking manual override switch ALSO sets disabled ( time code 00:05 in my video)
2. With manual override switch on, then selecting sleep mode switch to on (00:11) clears the manual override (00:12), resulting in unit not sleeping (00:13)
3. With manual override off, selecting sleep mode switch to on (00:21) sets unit to sleep (00:22), but the sleep mode switch turns itself off (00:23) (I know the video makes it look like I might have mouse clicked it off, but you will have to trust me that it flipped back to off by itself)

**Comment by fhteagle** [comment]
Hmm. I 100% agree that the switch should match the actual unit status in the end. Somehow what the integration is sending for commands ends up with confusing / unexpected / mismatching controls and indications though.

For 1: when user/an automation commands the manual override switch to on, can the integration look at the position of the sleep switch and decide whether to set status: active or status: disable in the override HTTP API command accordingly? 

For 2 and 3.  Same thing but in reverse, when sleep switch is user/automation commanded to switch: turn_on, look at the status of the manual override switch and send the correct corresponding command?

**Comment by firstof9** [comment]
Logic like that can be added, but if sleep is on, one should technically toggle sleep off if they want the charger to activate.

The manual override switch uses this function: https://openevse.stoplight.io/docs/openevse-wifi-v4/39daed23114f7-toggle-the-manual-override

**Comment by doppiaemme** [comment]
Am I finding confusing the Sleep name? 
Sleep in OpenEVSE is referenced as an advanced setting to disable the pilot signal for cars that don't support it well.

**Comment by firstof9** [comment]
> Am I finding confusing the Sleep name?

The name comes directly from the RAPI command for it:

<h2 style="border-bottom: 1px solid rgb(238, 238, 238); margin-top: 2px; padding-bottom: 0.2em; color: rgb(51, 51, 51); font-family: Helvetica, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">RAPI System Functions</h2>

Function | Description
-- | --
$FB | LCD Backlight Color (0-7)
$FD | Disable EVSE
$FE | Enable EVSE
$FP | Output text at x y position text to LCD (x y text)
$FR | Reset EVSE
$FS | Put EVSE to sleep
$FF | Enable/disable feature (feature_id 0\|1)Front panel (B)utton, (D)iode check, G(F)I self-test, (G)round check, Stuck (R)elay check, (T)emp monitoring, (V)ent required

**Comment by fhteagle** [comment]
> The manual override switch uses this function: https://openevse.stoplight.io/docs/openevse-wifi-v4/39daed23114f7-toggle-the-manual-override

Would it be better to use: https://openevse.stoplight.io/docs/openevse-wifi-v4/e0ab0a4ad5e1e-set-the-manual-override , setting the value of override that HA thinks it should be, instead of just toggle and hoping the result matches the user input?

**Comment by firstof9** [comment]
> instead of just toggle and hoping the result matches the user input?

Nope that's basically exactly what the switch is for, toggling nothing more. That's what switch entities in HA are for, toggle on/off.

Setting the manual override with specific parameters is via the service call.

**Comment by fhteagle** [comment]
HA has button entities and button press service calls availalbe, that make more sense to me to accomplish a toggle (i.e. I don't care what the state is, just change it). Switch, to me, should operate like a lightswitch. Flip it on, and it stays on, and if the lightbulb is working, it lights and stays lit. 

To reduce end user confusion, I strongly suggest pairing toggle with button, and set state with switch.

**Comment by firstof9** [comment]
Button entities are not toggles.

```
A button entity is an entity that can fire an event / trigger an action towards a device or service but 
remains stateless from the Home Assistant perspective. It can be compared to a real live 
momentary switch, push-button, or some other form of a stateless switch. It is, however, not 
suitable for implementing actual physical buttons; the sole purpose of a button entity is to provide a 
virtual button inside Home Assistant.
```

Switches are toggles.
```
A switch entity turns on or off something, for example a relay. To represent something which can be 
turned on or off but can't be controlled, for example a wall switch which transmits its state but can't 
be turned on or off from Home Assistant, a Binary Sensor is a better choice. To represent something 
which doesn't have a state, for example a door bell push button, a custom event or a Device Trigger is a better choice.
```

Since the override toggle function isn't a "momentary switch" or "stateless" it's a toggle.

**Comment by fhteagle** [comment]
Again, with all due respect, I disagree with your interpretation of manual override being stateless. 

If an override is set, response is the contents of the override:
$ curl --request GET   --url http://openevse.local/override   --header 'Content-Type: application/json'
{"charge_current":14,"auto_release":true}

If an override is not set:
$ curl --request GET \
  --url http://openevse.local/override \
  --header 'Content-Type: application/json'
{"msg":"No manual override"}

Therefore, override can have two states, "set/on" and "not set/off".

Same for sleeping, which is explicitly returned a property named _state_ .......

With override set, and sleeping true aka state disabled.
$ curl --request GET   --url http://openevse.local/override   --header 'Content-Type: application/json'
{"state":"disabled","charge_current":14,"auto_release":true}

Therefore there are three possible state combinations between the override itself and the "sleeping" state:
1.  override set and state active (aka not sleeping)
2.  override set and state disabled (aka sleeping)
3.  override not set, unit active by default but may be modified by claims instead

**Comment by firstof9** [comment]
> Therefore, override can have two states, "set/on" and "not set/off".

Correct which isn't "stateless" as I stated previously and is a toggle aka `Switch`, something like the RAPI command `$FR` to reset the EVSE would qualify for a `button` entity, since it's "stateless".

**Comment by fhteagle** [comment]
Okay, then if we agree that an override has a state, and sleeping definitely has states, then I believe the second block of text that you quoted in https://github.com/firstof9/openevse/issues/161#issuecomment-1474251591 supports my position rather than refutes it. My position is that the switch position of the HA control should represent the state of the override, and the state of the sleeping-ness of the unit, respectively. 

1. Override set and state: active returned in HTTP GET should result in HA Override switch on and HA sleeping switch off
2. Override set and state: disabled returned in HTTP GET should result in HA Override switch on and HA sleeping switch on
3. Override not set should result in HA Override switch off and HA sleeping switch off

Or, better yet, the HA switches could be removed entirely, these three possible states could all be combined with valid amps request options between 6 and unit configured max amps into a single HA select dropdown option set, but that idea probably should go in a new discussion or issue instead.

**Comment by firstof9** [comment]
> My position is that the switch position of the HA control should represent the state of the override

The switch reports what the unit is telling it, if the override is enabled, the switch is on, if the override is off, the switch is off.
The firmware is telling HA if the switch is on or off. The integration is only displaying exactly what the unit is telling it verbatim.

**Comment by fhteagle** [comment]
The "Sleep Mode" HA switch is likewise supposed to represent whether the unit is active and offering current to the vehicle (HA switch off / left position / not colored) vs disabled and not offering current to the vehicle (HA switch on / right position / colored), correct?  

If so, the sleep mode switch resulting status is definitely broken, see time codes 00:08 and 00:25 in my video above for evidence of this.

If not, there is a mismatch between an HA switch being a stateful representation of the current state of the unit (as HA override switch does) versus the sleep mode switch being used incorrectly like a momentary pushbutton type of control. 

If the two switches look the same, it is only reasonable for the end user to expect them to act the same and represent states the same way................................................

**Comment by firstof9** [comment]
Technically the "Sleep Mode" switch was initially to support the older firmware, but it's also based on the status of the device, when the state is `254` this is "Sleeping".

Again the state of the switches is completely based on what the unit is reporting. I understand that it maybe confusing, but the firmware is reporting the states.

**Comment by fhteagle** [comment]
> when the state is `254` this is "Sleeping".

I see that "state":254" vs "state: 3" vs "state: 2" is correctly reflected in sensor.*evse_charging_status . You are getting that integer value from in /status endpoint (formerly /ws endpoint, seems API docs are old regarding that), correct?

However, state: 254 vs state:3 or 2 in /status endpoint is definitely not getting reflected in the position of switch.*evse_sleep_mode at all. 

![Screenshot_20230317_142059](https://user-images.githubusercontent.com/9244278/226029215-c91ad5ae-7a2b-4d0a-aa8d-d3492b0762f5.png)

Is it possible this is because I have set a custom name for each of my units (1756evse  and 2691evse), so my entity switch is actually named switch.1756evse_sleep_mode, versus the stock entity name for the switch of switch.openevse_sleep_mode ...?

**Comment by firstof9** [comment]
Has nothing to do with the entity's name. If the state is `254` the sleeping switch turns on, any other status it is off.

In `2.1.14` I broke the sleep switch, it's been fixed in `2.1.15`.

**Comment by fhteagle** [comment]
Thanks for the fix, and I can confirm that with 2.1.15 the switch.*evse_sleep_mode switch position does seem to match the state integer in /status.
